README.md
# twitch_plays_souls

To see a live demo, click [here](https://clips.twitch.tv/CleanBrightCoffeePRChase-UJt_OTK-Dl8fXmv3).

## Important Notes
Disclaimer 1: **Please be aware that other people will be simulating keyboard presses to your computer. 
While I will do my part to ensure that no harmful scenarios arise,
PLEASE make sure to remove any revealing info from your desktop screen or other running programs if technical difficulties arise when live**

Disclaimer 2: **I have no idea if fromsoft tracks speed of key presses so please take necessary precautions to prevent any possible bans**

## Prerequisites
- Have Windows Powershell on your computer (type 'powershell' in your search bar and see if it shows up)
- HIGHLY recommend Microsoft Visual Studio Code with the Python extension to simplify code changes and execution
- Something to open up ZIP files with, like 7-Zip or WinRAR
- Dark Souls: Prepare to Die Edition preferably, otherwise the cash grab known as "Remastered"
    - Make the character beforehand, chat will not have full access to the keyboard to write the character name, and they will probably get stuck in advanced settings

## Requirements
- Install Python specifically 3.9.6. I am using it at this time on a Windows 10 64-bit computer. Can be found at Python website [here](https://www.python.org/downloads/)
    - After downloading, right click the installer to run it in administrator mode 
    - Choose the destination path to be in your main drive's Program Files(x86) path, for example: `C:\Program Files (x86)`
    - Make sure to choose to install 'pip', 'Documentation', 'tcl/tk and IDLE', and 'Python test suite' with the installer
    - In advanced settings of installer, select:
        - 'Install for all users'
        - 'Create shortcuts for installed applications'
        - 'Add Python to environment variables'
        - 'Precompile standard library'
- Install a Python library called 'pywin32'
    1. Open up Command Prompt (or Windows Powershell)
    2. Input the command below to install
    ```
    pip install pywin32
    ```
- Authentication token for Twitch connection
    - Generate an OAuth token with this website [here](https://twitchapps.com/tmi/)
    - KEEP THE TOKEN HANDY, YOU NEED TO EDIT A SCRIPT AND PASTE THE TOKEN

## Installation
- Download this script by clicking the green Code button on top, and 'Download as ZIP'
- Save the zip contents somewhere you can easily access it
- Open up `main.py` preferably with Microsoft VS Code, otherwise Notepad
- Paste the OAUTH Token from above in the line that starts with `token =`
- Replace `nickname` to your username
- Replace `channel` to your username but with a `#` in front and no space (i.e. `#mohomie`)
- Save your changes

## Preparing the game
- Boot up Dank Souls
- Go to the settings where you configure keybinds (we are NOT using a controller to keep this as simple as possible, but you can have it connected too)
- Change the controls to the following... TODO

## Execution
- If using Microsoft VS Code to make the replacements above:
    - press CTRL+` to open up a terminal at the bottom of your screen
    - In that terminal just type `python .\main.py`
    - You'll see some print messages but if the last couple of lines say NEW MESSAGE in them, it should be working.
    - Click into the window where DS1 is to focus onto the game
    - Go for it!

## Twitch Limitations on Chat
Note: These are from my observations, probably will need a live test to determine any more issues
- A person is unable to type the same exact message twice in a row.
    - Possible fix may be adding alternatives to each command
- A person spamming too many different messages in a short span of time may cause them to be shadowbanned or timeout based on Twitch's global rules. 
    - Possible fix might be to enable slow mode of 3 seconds.

## TODO
[X] Add DS Remastered control support

[X] Perform stress testing by simulating a rowdy twitch chat and seeing how program holds up
    - Can handle random commands generated every 1/100th of a second
    - Does not execute all commands due to character being locked in animations

[X] Add some sort of failsafe command
    - **From another device**, type STOP **in all caps** into chat to terminate the program

[X] Validate that above readme works
    - Tested installation settings with downloading ZIP and installing pywin32 only, runs successfully

[X] Modify controls to use letter keys only
    - Already implemented

[X] Perform long term test

[ ] Detect when DS closes to trigger auto shutdown

[X] Add additional commands to START and STOP the stream


changelog.md
# Changelog

All changes can be found here

## 1.1.0 - 7/20/21
- Added new commands:
    - 'toggleblock' - either holds block or releases block
    - 'togglesprint' - holds down circle so any movement is a sprint in movement direction.
        - cancelled either by typing 'togglesprint' again or by typing 'o'/'circle'
- Lowered delay between commands from 0.2 seconds to 0.1 seconds
- Fixed minor bug where unintended command is pressed since first letter of last word was considered a command attempt

keyboard.py
import ctypes
import time

SendInput = ctypes.windll.user32.SendInput

# C struct redefinitions 
PUL = ctypes.POINTER(ctypes.c_ulong)
class KeyBdInput(ctypes.Structure):
    _fields_ = [("wVk", ctypes.c_ushort),
                ("wScan", ctypes.c_ushort),
                ("dwFlags", ctypes.c_ulong),
                ("time", ctypes.c_ulong),
                ("dwExtraInfo", PUL)]

class HardwareInput(ctypes.Structure):
    _fields_ = [("uMsg", ctypes.c_ulong),
                ("wParamL", ctypes.c_short),
                ("wParamH", ctypes.c_ushort)]

class MouseInput(ctypes.Structure):
    _fields_ = [("dx", ctypes.c_long),
                ("dy", ctypes.c_long),
                ("mouseData", ctypes.c_ulong),
                ("dwFlags", ctypes.c_ulong),
                ("time",ctypes.c_ulong),
                ("dwExtraInfo", PUL)]

class Input_I(ctypes.Union):
    _fields_ = [("ki", KeyBdInput),
                 ("mi", MouseInput),
                 ("hi", HardwareInput)]

class Input(ctypes.Structure):
    _fields_ = [("type", ctypes.c_ulong),
                ("ii", Input_I)]

# Actuals Functions

def PressKey(hexKeyCode):
    extra = ctypes.c_ulong(0)
    ii_ = Input_I()
    ii_.ki = KeyBdInput( 0, hexKeyCode, 0x0008, 0, ctypes.pointer(extra) )
    x = Input( ctypes.c_ulong(1), ii_ )
    ctypes.windll.user32.SendInput(1, ctypes.pointer(x), ctypes.sizeof(x))

def ReleaseKey(hexKeyCode):
    extra = ctypes.c_ulong(0)
    ii_ = Input_I()
    ii_.ki = KeyBdInput( 0, hexKeyCode, 0x0008 | 0x0002, 0, ctypes.pointer(extra) )
    x = Input( ctypes.c_ulong(1), ii_ )
    ctypes.windll.user32.SendInput(1, ctypes.pointer(x), ctypes.sizeof(x))

# directx scan codes http://www.gamespp.com/directx/directInputKeyboardScanCodes.html
# while (True):
#     PressKey(0x11)
#     time.sleep(0.1)
#     ReleaseKey(0x11)
#     time.sleep(1)

generator.py
import random
import pyautogui
import time

ds1_keymap = {
    'up': 0x11, # w
    'left': 0x1E, # a
    'down': 0x1F, # s
    'right': 0x20, # d
    'dup': 0xC8 + 1024, #arrowup
    'dleft': 0xCB + 1024, #arrowleft
    'ddown': 0xD0 + 1024, #arrowdown
    'dright': 0xCD + 1024, #arrowright
    'cup': 0x17, # i
    'cleft': 0x24, # j
    'cdown': 0x25, # k
    'cright': 0x26, # l
    'r3': 0x18, # o
    'l1': 0x2C, # z
    'l2': 0x2D, # x
    'r1': 0x23, # h
    'r2': 0x16, # u
    's': 0x12, # e
    'sq': 0x12, # e
    'square': 0x12, # e
    't': 0x2E, # c
    'tri': 0x2E, # c
    'triangle': 0x2E, # c
    'o': 0x2F, # v
    'circle': 0x2F, # v
    'x': 0x30, # b
    'start': 0x31, # n
    'select': 0x22 # g
}

keys = list(ds1_keymap.keys())
print(keys)

if __name__ == '__main__':
    while True:
        key = random.choice(keys)
        pyautogui.write(key)
        pyautogui.press('enter')
        time.sleep(1)

main.py
import sys
import re
import socket
import logging
import time
import keyboard
import win32api
import win32con

# Full list of character mappings can be found at https://raw.githubusercontent.com/SerpentAI/SerpentAI/dev/serpent/input_controllers/native_win32_input_controller.py
# ds1_keymap = {
#     'forward': 0x11, # w
#     'left': 0x1E, # a
#     'back': 0x1F, # s
#     'right': 0x20, # d
#     'dup': 0xC8 + 1024, #arrowup
#     'dleft': 0xCB + 1024, #arrowleft
#     'ddown': 0xD0 + 1024, #arrowdown
#     'dright': 0xCD + 1024, #arrowright
#     'cup': 0x17, # i
#     'cleft': 0x24, # j
#     'cdown': 0x25, # k
#     'cright': 0x26, # l
#     'r3': 0x18, # o
#     'l1': 0x2C, # z
#     'l2': 0x2D, # x
#     'r1': 0x23, # h
#     'r2': 0x16, # u
#     'sq': 0x12, # e
#     'square': 0x12, # e
#     'tri': 0x2E, # c
#     'triangle': 0x2E, # c
#     'o': 0x2F, # v
#     'circle': 0x2F, # v
#     'x': 0x30, # b
#     'start': 0x31, # n
#     'select': 0x22 # g
# }

ds1_keymap = {
    'forward': 0x11, # w
    'left': 0x1E, # a
    'back': 0x1F, # s
    'right': 0x20, # d
    'dup': 0x1B, # ]      # switch magic ]
    'dleft': 0x0C, # -    # switch left -
    'ddown': 0X1A, # [    # switch item [
    'dright': 0x0D, # =   # switch right weapon =
    'menuup': 0x32, # m      # select up m
    'mup': 0x32, # m      # select up m
    'menuleft': 0x21, # f    # select left f
    'mleft': 0x21, # f    # select left f
    'menudown': 0x15, # y    # select down y
    'mdown': 0x15, # y    # select down y
    'menuright': 0x19, # p    # select right p
    'mright': 0x19, # p    # select right p
    'camup': 0x17, # i
    'cup': 0x17, # i
    'camleft': 0x24, # j
    'cleft': 0x24, # j
    'camdown': 0x25, # k
    'cdown': 0x25, # k
    'camright': 0x26, # l
    'cright': 0x26, # l
    'r3': 0x18, # o
    'l1': 0x2C, # z
    'l2': 0x2D, # x
    'r1': 0x23, # h
    'r2': 0x16, # u
    'sq': 0x12, # e
    'square': 0x12, # e
    'tri': 0x2E, # c
    'triangle': 0x2E, # c
    'o': 0x2F, # v
    'circle': 0x2F, # v
    'x': 0x30, # b
    'enter': 0x30, # b
    'start': 0x31, # n
    'select': 0x22 # g
}

ACTIVE = False
blocking = False # flag used for determining if character holding block or not
sprinting = False # flag used for determining if character holding sprint or not

dodgeSet = {'dodge', 'dodgeleft', 'dodgeright', 'dodgeback'}

def dodge(key):
    global sprinting
    delay = 0.1
    if sprinting:
        keyboard.ReleaseKey(ds1_keymap['o'])
        time.sleep(delay)
    if key == '':
        keyboard.PressKey(ds1_keymap['forward'])
        keyboard.PressKey(ds1_keymap['circle'])
        time.sleep(delay)
        keyboard.ReleaseKey(ds1_keymap['forward'])
        keyboard.ReleaseKey(ds1_keymap['circle'])
    else:
        keyboard.PressKey(ds1_keymap[key])
        keyboard.PressKey(ds1_keymap['circle'])
        time.sleep(delay)
        keyboard.ReleaseKey(ds1_keymap[key])
        keyboard.ReleaseKey(ds1_keymap['circle'])
    if sprinting:
        time.sleep(delay)
        keyboard.PressKey(ds1_keymap['o'])

def press(key):
    global blocking
    global sprinting
    key = key.lower()
    if key == 'l1' and blocking: return
    delay = 0.1
    # If input command is to move or block, set a higher duration for action
    if key == 'forward' or key == 'left' or key == 'back' or key == 'right' or key == 'l1':
        delay = 0.5
    elif key == 'toggleblock':
        if blocking: 
            keyboard.ReleaseKey(ds1_keymap['l1'])
            blocking = False
        else:
            keyboard.PressKey(ds1_keymap['l1'])
            blocking = True
        return
    elif key == 'togglesprint':
        if sprinting:
            keyboard.ReleaseKey(ds1_keymap['o'])
            sprinting = False
        else:
            keyboard.PressKey(ds1_keymap['start'])
            time.sleep(0.1)
            keyboard.ReleaseKey(ds1_keymap['start'])
            time.sleep(0.1)
            keyboard.PressKey(ds1_keymap['o'])
            sprinting = True
        return
    elif key in dodgeSet:
        if key[:5] == 'dodge':
            dir = key[5:] # extract dodge direction
            dodge(dir)
    direct = ds1_keymap.setdefault(key,'')
    print(direct)
    if direct != '':
        keyboard.PressKey(direct)
        time.sleep(delay)
        keyboard.ReleaseKey(direct)

# Function to detect clicks specifically. Not used currently but can be used for other games that require mouse input
def leftClick():
    win32api.mouse_event(win32con.MOUSEEVENTF_LEFTDOWN,0,0)
    time.sleep(.1)
    win32api.mouse_event(win32con.MOUSEEVENTF_LEFTUP,0,0)
    print('Left Click')


if __name__ == '__main__':
    print("Initializing Twitch Connection...")
    sock = socket.socket()
    logging.basicConfig(level=logging.DEBUG,
                        format='%(asctime)s — %(message)s',
                        datefmt='%Y-%m-%d_%H:%M:%S',
                        handlers=[logging.FileHandler('chat.log', encoding='utf-8')])
    
    print("Initializing Parameters...")
    server = 'irc.chat.twitch.tv'
    port = 6667
    nickname = 'mohomie' # rename this to streamer name
    token = 'oauth:5bcb60h4009zv85rqbh09p0vqmellz' # replace with your own token using link here... https://twitchapps.com/tmi/
    channel = '#mohomie' # rename this to streamer name with a '#' in front of it

    sock.connect((server, port))

    sock.send(f"PASS {token}\n".encode('utf-8'))
    sock.send(f"NICK {nickname}\n".encode('utf-8'))
    sock.send(f"JOIN {channel}\n".encode('utf-8'))

    print("Listening...")
    while True:
        resp = sock.recv(512).decode('utf-8') # chat single message limit is 500
        if len(resp) > 0:
            # Reply to a PING with a PONG to ensure we don't time out
            print("NEW MESSAGE: " + resp)
            if resp.startswith('PING'):
                sock.send("PONG\n".encode('utf-8'))
            else:
                key = resp.split()[-1]
                if key[0] == ':': 
                    key = key[1:]
                    if key == 'EXIT' and (resp.startswith(':napstarf!') or resp.startswith(':mohomie!')):
                        print("-----------------ADMIN STOP---------------")
                        sys.exit()
                    elif key == 'PAUSE' and (resp.startswith(':napstarf!') or resp.startswith(':mohomie!')):
                        print("-----------------ADMIN PAUSE---------------")
                        ACTIVE = False
                    elif key == 'START' and (resp.startswith(':napstarf!') or resp.startswith(':mohomie!')):
                        print("-----------------ADMIN START---------------")
                        ACTIVE = True
                    # Hardcode pressing the 'Enter' key
                    elif key.startswith('ENTER') and (resp.startswith(':napstarf!') or resp.startswith(':mohomie!')):
                        print("-----------------ADMIN ENTER---------------")
                        keyboard.PressKey('0x1C')
                        time.sleep(0.1)
                        keyboard.ReleaseKey('0x1C')
                    else:
                        
                        if ACTIVE: 
                            print('ATTEMPTING TO PRESS ' + key)
                            press(key)
                        else: print("CHAT COMMANDS CURRENTLY PAUSED")

# NEW MESSAGE: :lokoheimer!lokoheimer@lokoheimer.tmi.twitch.tv PRIVMSG #mohomie :cleft